### ISSUE TITLE
Report unnecessary comparisons between enums and different classes.

### ISSUE THREAD

**Issue by aDotInTheVoid** [microsoft/pyright#9228]
**Is your feature request related to a problem? Please describe.**

The following code is buggy, as `b: Bar` will never be `Foo.A`, so the `print("hiya")` case can't be reached.

```python
# pyright: strict
import enum
import typing


@typing.final
class Foo(enum.Enum):
    A = "A"


@typing.final
class Bar:
    pass


def demo(b: Bar):
    match b:
        case Foo.A:
            print("hiya")
        case Bar():
            print("hey")
```

It would be nice if pyright could catch this, but currently it doesn't:

```
% pyright ./mcve.py
0 errors, 0 warnings, 0 informations 
```

**Describe the solution you’d like**

The reportUnnecessaryComparison diagnostic should fire here. 

**VS Code extension or command-line**

pyright 1.1.384 cli

**Comment by erictraut** [comment]
The challenge here is that value pattern matching uses simple `__eq__` tests, and a static type checker cannot generally determine whether `__eq__` will always fail for a given value. That's the nature of Python, which allows overrides of most operators. In this case, for example, the expression `Foo.A.__eq__` is a method that accepts any `object` and returns a `bool` value, so there's no guarantee (from the perspective of a type checker) that it will return `False` if it is passed a `Bar` instance. In short, I don't think it's possible to implement the check that you're requesting here without generating false positive errors in some cases.
